In [ ]:
!pip install langchain groq
!pip install langchain-groq groq



     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 668.7 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 997.8/997.8 kB 18.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.3/106.3 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 394.9/394.9 kB 20.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.5/150.5 kB 9.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 141.9/141.9 kB 9.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 4.0 MB/s eta 0:00:00
  Attempting uninstall: tenacity
    Found existing installation: tenacity 9.0.0
    Uninstalling tenacity-9.0.0:
      Successfully uninstalled tenacity-9.0.0


In [ ]:
from langchain_groq import ChatGroq
from langchain.chains import LLMChain, SimpleSequentialChain
from langchain_core.messages import AIMessage, HumanMessage, SystemMessage
from langchain.prompts import Prompt
from langchain import PromptTemplate, LLMChain
import os

# Set your Groq API key
os.environ["GROQ_API_KEY"] = "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"

# Initialize the Groq LLM
llm = ChatGroq(model_name="mixtral-8x7b-32768")



Question 1: Chain-of-Thought Prompting Implement a Chain-of-Thought prompting system to solve word problems. Your system should break down the problem-solving process into steps and show its reasoning.

Task:

● Create a function that takes a word problem as input.

● Use the OpenAI API or any other LLM API to generate a step-by-step solution.

● The output should show each step of the reasoning process.

● Test your function with at least two word problems of varying complexity.

Example word problem: "A baker has 150 eggs. If each cake requires 3 eggs and each cookie requires 2 eggs, how many cakes and cookies can the baker make if they want to use all the eggs and make an equal number of each?"

In [ ]:
#Setup HumanMessage, SystemMessage for simple prompt.
example = """
Question: A gardener has 36 seeds. Each plant requires either 3 seeds for a rose or 4 seeds for a tulip. The gardener wants to use all the seeds and plant an equal number of roses and tulips.Answer: 10.
Answer: 5.14
Question: A library has 45 books. If they lent out 12 books and then received 8 more as donations, how many books does the library have now?
Answer:
"""
human_message = HumanMessage(content=example)
system_message = SystemMessage(content="You are helpful assistant.")


new_prompt = (system_message + human_message)

print (f"Prompt: {new_prompt}")

chain = LLMChain(llm=llm, prompt=new_prompt)
reponse = chain.run({})
print(f"Response: {reponse}")

Prompt: input_variables=[] messages=[SystemMessage(content='You are helpful assistant.'), HumanMessage(content='\nQuestion: A gardener has 36 seeds. Each plant requires either 3 seeds for a rose or 4 seeds for a tulip. The gardener wants to use all the seeds and plant an equal number of roses and tulips.Answer: 10.\nAnswer: 5.14\nQuestion: A library has 45 books. If they lent out 12 books and then received 8 more as donations, how many books does the library have now?\nAnswer:\n')]


/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:151: LangChainDeprecationWarning: The class `LLMChain` was deprecated in LangChain 0.1.17 and will be removed in 1.0. Use RunnableSequence, e.g., `prompt | llm` instead.
  warn_deprecated(
/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:151: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use invoke instead.
  warn_deprecated(


Response: The library has 41 books now.

Here's the reasoning:
1. The library starts with 45 books.
2. They lend out 12 books, leaving them with 45 - 12 = 33 books.
3. Then, they receive 8 books as donations, adding to their collection.
4. So, the total number of books in the library now is 33 + 8 = 41 books.


Question 2: Prompt-chaining Develop a prompt-chaining system that generates a short story based on user input. Thesystem should use multiple prompts in sequence, with each prompt building on the output of theprevious one.

Task:

● Create a series of at least 3 prompts that will be chained together.

● The first prompt should ask for a genre and main character.

● Subsequent prompts should build on previous outputs to develop the plot and conclusion.

● Implement the chaining logic to pass information between prompts.

● Allow the user to input the initial genre and main character.

● Output the final story generated through the chain.

In [ ]:
# System message setup
system_message = SystemMessage(content="You are a creative assistant for generating short stories.")

# Prompt 1: Get genre and main character
initial_prompt = HumanMessage(content="Provide a genre and a main character for a short story.")

# Create the chain and get the genre and character
chain1 = LLMChain(llm=llm, prompt=system_message + initial_prompt)
response1 = chain1.run({})

try:
    genre, character = response1.split(', ')
except ValueError as e:
    print(f"Error parsing response: {e}")
    genre, character = None, None

# Example response
response1 = "Fantasy, The Lost City"
print(f"Genre: {genre}, \nCharacter: {character}")

# Prompt 2: Develop the plot based on the genre and character
plot_prompt = HumanMessage(content=f"Based on the genre '{genre}' and the main character '{character}', write a brief plot outline for a short story.")

# Create the chain and get the plot
chain2 = LLMChain(llm=llm, prompt=system_message + plot_prompt)
plot = chain2.run({})

# Prompt 3: Write the conclusion based on the plot
conclusion_prompt = HumanMessage(content=f"Based on the plot outline provided, write the conclusion for the short story.")

# Create the chain and get the conclusion
chain3 = LLMChain(llm=llm, prompt=system_message + conclusion_prompt)
conclusion = chain3.run({})

# Combine all parts to form the final story
final_story = f"Genre: {genre}\nMain Character: {character}\nPlot: {plot}\nConclusion: {conclusion}"
print(final_story)


Error parsing response: not enough values to unpack (expected 2, got 1)
Genre: None, 
Character: None
Genre: None
Main Character: None
Plot: Title: The Invisible Realm

In a world where the invisible realm coexists with the visible one, our protagonist, an unseen entity, finds themselves lost and lonely. Unable to communicate with the physical world, they watch as humans go about their lives, longing for connection.

One day, they stumble upon a young girl, Lily, who has the unique ability to perceive unseen entities. A friendship blossoms between them, and the invisible entity finally finds the companionship they've yearned for.

However, this newfound friendship draws the attention of a malevolent force in the invisible realm. The entity must now protect Lily and her gift from those who wish to exploit it. In the process, they discover their own strength and purpose, forever changing the relationship between the two realms.
Conclusion: Plot Outline:
In a small, tight-knit town, a mys

Question 3: Tree-of-Thought Prompting Implement a Tree-of-Thought prompting system to solve a complex problem. The system should explore multiple paths of reasoning and select the most promising one. Task:

● Create a function that takes a complex problem as input.

● Generate at least 3 different initial approaches to solving the problem.

● For each approach, develop 2-3 possible next steps.

● Implement a method to evaluate the promise of each path.

● Select the most promising path and generate a final solution.

● Test your system with the following problem:

Problem: "Design a sustainable urban transportation system for a city of 1 million people. Consider factors such as environmental impact, cost-effectiveness, and accessibility for all residents."

In [ ]:
# First prompt template
first_prompt = PromptTemplate(
    input_variables=["topic"],
    template="Write a short introduction about {topic}."
)

# Second prompt template
second_prompt = PromptTemplate(
    input_variables=["introduction"],
    template="Based on this introduction, suggest 3 key points to expand upon: {introduction}"
)

# Third prompt template
third_prompt = PromptTemplate(
    input_variables=["key_points"],
    template="For each of these key points, provide a brief explanation: {key_points}"
)

# Create chains
chain_one = LLMChain(llm=llm, prompt=first_prompt)
chain_two = LLMChain(llm=llm, prompt=second_prompt)
chain_three = LLMChain(llm=llm, prompt=third_prompt)

# Create the sequential chain
overall_chain = SimpleSequentialChain(
    chains=[chain_one, chain_two, chain_three],
    verbose=True
)

# Run the chain
topic = "Design a sustainable urban transportation system for a city of 1 million people.Consider factors such as environmental impact, cost-effectiveness, and accessibility for allresidents"
result = overall_chain.run(topic)

print(result)



> Entering new SimpleSequentialChain chain...
Designing a sustainable urban transportation system for a city of 1 million people is a complex task that requires careful consideration of various factors. The goal is to create a system that minimizes environmental impact, is cost-effective, and accessible to all residents.

To achieve these objectives, a multi-modal transportation system that includes a mix of public transit, active transportation (such as walking and cycling), and shared mobility options (like car-sharing and ride-hailing) should be prioritized. Public transit, in particular, should be the backbone of the system, with frequent and reliable service that connects all parts of the city.

In terms of environmental impact, prioritizing electric or low-emission vehicles for public transit and shared mobility options is crucial. Additionally, designing the system to encourage active transportation can help reduce greenhouse gas emissions and improve air quality.

Cost-effect